tools_lib/src/stack_collapse.rs

```diff
@@ -1,183 +1,215 @@
-                timestamp: SystemTime::UNIX_EPOCH + Duration::from_secs(1),
-                timestamp_kind: TimestampKind::Start,
-                thread_id: 1,
-            },
-            Event {
-                event_kind: "Query".into(),
-                label: "EventB".into(),
-                additional_data: &[],
-                timestamp: SystemTime::UNIX_EPOCH + Duration::from_secs(3),
-                timestamp_kind: TimestampKind::Start,
-                thread_id: 2,
-            },
-            Event {
-                event_kind: "Query".into(),
-                label: "EventA".into(),
-                additional_data: &[],
-                timestamp: SystemTime::UNIX_EPOCH + Duration::from_secs(2),
-                timestamp_kind: TimestampKind::End,
-                thread_id: 1,
-            },
-            Event {
-                event_kind: "Query".into(),
-                label: "EventA".into(),
-                additional_data: &[],
-                timestamp: SystemTime::UNIX_EPOCH + Duration::from_secs(4),
-                timestamp_kind: TimestampKind::Start,
-                thread_id: 2,
-            },
-            Event {
-                event_kind: "Query".into(),
-                label: "EventA".into(),
-                additional_data: &[],
-                timestamp: SystemTime::UNIX_EPOCH + Duration::from_secs(5),
-                timestamp_kind: TimestampKind::End,
-                thread_id: 2,
-            },
-            Event {
-                event_kind: "Query".into(),
-                label: "EventB".into(),
-                additional_data: &[],
-                timestamp: SystemTime::UNIX_EPOCH + Duration::from_secs(6),
-                timestamp_kind: TimestampKind::End,
-                thread_id: 2,
-            },
-        ];
-
-        let recorded_stacks = super::collapse_stacks(events.iter().cloned(), 1000);
-
-        let mut expected_stacks = HashMap::<String, usize>::new();
-        expected_stacks.insert("rustc;EventB;EventA".into(), 1);
-        expected_stacks.insert("rustc;EventB".into(), 2);
-        expected_stacks.insert("rustc;EventA".into(), 1);
+        let mut b = ProfilingDataBuilder::new();
+
+        //                       <--------->
+        //   <------------><-------------------->
+        //   100           300   400       600   700
+
+        b.interval("Query", "EventA", 0, 100, 300, |_| {});
+        b.interval("Query", "EventB", 0, 300, 700, |b| {
+            b.interval("Query", "EventA", 0, 400, 600, |_| {});
+        });
+
+        let profiling_data = b.into_profiling_data();
+
+        let recorded_stacks = super::collapse_stacks(&profiling_data, Duration::from_nanos(1));
+
+        let mut expected_stacks = HashMap::<String, u64>::new();
+        expected_stacks.insert("rustc;EventB;EventA".into(), 200);
+        expected_stacks.insert("rustc;EventB".into(), 200);
+        expected_stacks.insert("rustc;EventA".into(), 200);
```

Comment:
Still need to rebase this onto https://github.com/rust-lang/measureme/pull/78...

Reply:
you can take it from michaelwoerister#1

Reply:
Took the whole algorithm from michaelwoerister#1 `:D`